---
title: "Integration v3.3.2 Analysis Results"
author: "Nate Olson"
date: "9/10/2018"
output: html_document
---

```{r echo = FALSE, message = FALSE}
library(drake)
library(tidyverse)
```

## Objective
Generate summary tables and figures for integration manuscript.



## High confidence region coverage and variants by chromosome

```{r}
loadd(hcr_cov_df)
loadd(hc_var_df)
loadd(hh_var_df)
```


```{r}
## Can add ti/tv, het/hom, and other stats from vcfstats
hc_vcf_tidy <- hc_vcf_df %>% 
    filter(Key %in% c("SNPs","Insertions","Deletions")) %>% 
    mutate(chrom = if_else(chrom == "chr1-22", "genome", chrom),
           Value = as.integer(Value)) %>% 
    spread(Key, Value) %>% 
    mutate(Indels = Insertions + Deletions)

hh_vcf_tidy <- hh_vcf_df %>% 
    filter(Key %in% c("SNPs","Insertions","Deletions")) %>% 
    mutate(Key = paste0("hh", Key)) %>% 
    mutate(chrom = if_else(chrom == "chr1-22", "genome", chrom),
           Value = as.integer(Value)) %>% 
    spread(Key, Value) %>% 
    mutate(hhIndels = hhInsertions + hhDeletions)


metrics_df <- left_join(hcr_cov_df, hc_vcf_tidy) %>% 
    left_join(hh_vcf_tidy)
```



__Table__ Metric summary for full genome
```{r}
metrics_df %>% filter(chrom == "genome") %>% 
    select(hgref, chrom, coverage_nonN, SNPs, hhSNPs, Indels, hhIndels)
```


```{r}
metrics_df %>% filter(chrom != "genome") %>% 
    ggplot() + 
    geom_point(aes(x = chrom, y = coverage_nonN, color = hgref)) + 
    theme_bw()
```

```{r}
metrics_df %>% filter(chrom != "genome") %>% 
    ggplot() + 
    geom_point(aes(x = chrom, y = SNPs, color = hgref)) + 
    theme_bw()
```

Figures comparing SNPs and Indels per chrom overall and for high-confidence


## Han Chinese Trio Mendelian inheritance analysis
```{r}
loadd(hc_trioincon_df)
loadd(hc_denovo_df)
violations_total <- nrow(hc_trioincon_df)
violations_snps <- hc_trioincon_df %>% filter(snp == TRUE) %>% nrow()
violations_indels <- hc_trioincon_df %>% filter(indel == TRUE) %>% nrow()

denovo_total <- nrow(hc_denovo_df)
denovo_snps <- hc_denovo_df %>% filter(snp == TRUE) %>% nrow()
denovo_indels <- hc_denovo_df %>% filter(indel == TRUE) %>% nrow()
```

__Mendelian Violations__   

Total: `r violations_total`   
SNPs: `r violations_snps`  
Indels: `r violations_indels`   

__De novo/ germ-line somatic mutations__  

Total: `r denovo_total`   
SNPs: `r denovo_snps`  
Indels: `r denovo_indels`   

### Manual curation results


## Benchmarking Results
precisionFDA Analysis Note
https://precision.fda.gov/notes/362-gatk4-giab-aj-and-asian-trio-benchmarking 

### Updated supplemental table 2
```{r}
loadd(benchmark_summary_tbl)
```


```{r}
supplemental_table <- benchmark_summary_tbl %>% 
    arrange(Type, Subset) %>% 
    mutate(Subset = if_else(Subset == "*", "all", Subset)) %>% 
    mutate(one_minus_recall = (1 - METRIC.Recall) * 100, 
           one_minus_precision = (1 - METRIC.Precision) * 100) %>% 
    dplyr::select(hgref, Type, Subset, one_minus_recall, one_minus_precision,  
           METRIC.Recall,  METRIC.Precision, METRIC.Frac_NA)

colnames(supplemental_table) <- c("Genome", "Type", "Subset", 
                                 "100% - recall", "100% - precision", 
                                 "Recall", "Precision", 
                                 "Fraction of calls outside high-conf bed")

```

Saving to googlesheet for integration paper
```{r}
library(googlesheets)
int_gs <- gs_title("integrationV3.3.2_supplemental_table2")
gs_edit_cells(ss = int_gs, ws = 1, input = supplemental_table,col_names = TRUE)
```

## System Information
Reproducible anaylsis info: 
```{r}
s_info <- devtools::session_info()
print(s_info$platform)
s_info$packages %>% filter(`*` == "*") %>% select(-`*`) %>% 
      knitr::kable()
```

software versions (bedtools, bgzip, tabix, rtg)
```{r}
system("bedtools --version",intern = TRUE)
```
```{r}
system("bgzip --version",intern = TRUE)
```

```{r}
system("tabix --version",intern = TRUE)
```

```{r}
system("~/sw/rtg-core-non-commercial-3.9.1/rtg --version",intern = TRUE)[1:2]
```

