---
title: "Integration v3.3.2 Analysis Results"
author: "Nate Olson"
date: '`r Sys.Date()`'
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

```{r echo = FALSE, message = FALSE}
library(drake)
library(tidyverse)
library(googlesheets)
```

Code to generate figures and tables for the GIAB Integration V3.3.2 manuscript. 


# Genome Metrics
Summary table and figures for the seven GIAB genomes. 

```{r message = FALSE}
## Loading and tidying data
loadd(hcr_cov_df)
loadd(hc_var_df)
loadd(hh_var_df)
loadd(hc_coding_df)

## Can add ti/tv, het/hom, and other stats from vcfstats
hc_vcf_tidy <- hc_vcf_df %>% 
    filter(Key %in% c("SNPs","Insertions","Deletions")) %>% 
    mutate(chrom = if_else(chrom == "chr1-22", "genome", chrom),
           Value = as.integer(Value)) %>% 
    spread(Key, Value) %>% 
    mutate(Indels = Insertions + Deletions)

hh_vcf_tidy <- hh_vcf_df %>% 
    filter(Key %in% c("SNPs","Insertions","Deletions")) %>% 
    mutate(Key = paste0("hh", Key)) %>% 
    mutate(chrom = if_else(chrom == "chr1-22", "genome", chrom),
           Value = as.integer(Value)) %>% 
    spread(Key, Value) %>% 
    mutate(hhIndels = hhInsertions + hhDeletions)


metrics_df <- left_join(hcr_cov_df, hc_vcf_tidy) %>% 
    left_join(hh_vcf_tidy)

## Combining metrics for summary table
highconf_cov <- hcr_cov_df %>% 
    filter(chrom == "genome") %>% 
    dplyr::select(hgref, nbases, coverage_nonN) 

highconf_calls <- hh_vcf_df %>% filter(Key %in% c("Passed Filters",
                                "SNPs","Insertions","Deletions", 
                                "Indels", "Phased Genotypes", 
                                "SNP Transitions/Transversions")) %>% 
    filter(chrom == "chr1-22") %>% 
    mutate(Value = str_remove(Value, " \\(.*")) %>% 
    spread(Key, Value) %>% 
    dplyr::rename(total_calls = `Passed Filters`)

calls_in_highconf <- hh_vcf_df %>% 
    filter(Key %in% "Passed Filters") %>% 
    filter(chrom == "chr1-22") %>% 
    spread(Key, Value) %>% 
    dplyr::rename(total_calls_in_highconf = `Passed Filters`)


coding_frac <- hc_coding_df %>% 
    filter(chrom == "genome") %>% 
    dplyr::select(hgref, frac_coding)

highconf_summary_metrics <- left_join(highconf_cov, coding_frac) %>%
    left_join(calls_in_highconf) %>% left_join(highconf_calls)

## Saving values to googlesheet
int_gs <- gs_title("integrationV3.3.2_supplemental")
gs_edit_cells(ss = int_gs, ws = 1, 
              input = highconf_summary_metrics, 
              col_names = TRUE)
```

## Metric summary table for full genomes HG001 - HG007 GRCh37. 
```{r}
highconf_summary_metrics %>% 
    dplyr::select(hgref, nbases, coverage_nonN, frac_coding, total_calls_in_highconf, 
                  total_calls, SNPs, Insertions, Deletions, Indels, 
                  `SNP Transitions/Transversions`, `Phased Genotypes`) %>% 
    mutate(coverage_nonN = round(coverage_nonN, 2)) %>% 
    mutate(frac_coding = round(frac_coding, 2)) %>% 
    gather("key", "value", -hgref) %>% 
    spread(hgref, value) %>% 
    knitr::kable()
```

__Table Value Descriptions__  

- Number of bases in highconf regions (chromosomes 1-22) (nbases)    
- Fraction of non-N bases covered in chromosomes 1-22) (coverage_nonN)  
- Fraction of RefSeq coding sequence covered (frac_coding)  
- Total number of calls in highconf regions (total_calls_in_highconf)  
- SNPs (from rtg vcfstats)  
- Insertions (from rtg vcfstats)  
- Deletions (from rtg vcfstats)  
- Block Substitutions (Indels - from rtg vcfstats)  
- Transition/Transversion ratio for SNVs (from rtg vcfstats)  
- Fraction phased globally or locally (from rtg vcfstats)  

## Chromosome Coverage Figure
```{r}
metrics_df %>% 
    filter(chrom != "genome", chrom != "chrX") %>% 
    mutate(chrom = factor(chrom, levels = paste0("chr", 1:22))) %>% 
    ggplot() + 
    geom_point(aes(x = chrom, y = coverage_nonN, color = hgref)) + 
    theme_bw() + 
    labs(x = "Chromsome", y = "Coverage of non-N Bases", color = "NIST ID") + 
    theme(axis.text.x = element_text(angle = -45, hjust = 0))
```


# Han Chinese Trio Mendelian inheritance analysis
```{r}
loadd(hc_trioincon_df)
loadd(hc_denovo_df)
violations_total <- nrow(hc_trioincon_df)
violations_snps <- hc_trioincon_df %>% filter(snp == TRUE) %>% nrow()
violations_indels <- hc_trioincon_df %>% filter(indel == TRUE) %>% nrow()

denovo_total <- nrow(hc_denovo_df)
denovo_snps <- hc_denovo_df %>% filter(snp == TRUE) %>% nrow()
denovo_indels <- hc_denovo_df %>% filter(indel == TRUE) %>% nrow()
```

__Mendelian Violations__   

Total: `r violations_total`   
SNPs: `r violations_snps`  
Indels: `r violations_indels`   

__De novo/ germ-line somatic mutations__  

Total: `r denovo_total`   
SNPs: `r denovo_snps`  
Indels: `r denovo_indels`   

## Manual curation results
1. Select 10 random variants (5 SNPs and 5 indels). Randomly selected variants from different chromosomes. 
```{r}
denovo_df <- hc_denovo_df %>% 
    mutate(var_type = if_else(indel == TRUE, "indel", "snp")) %>% 
    dplyr::select(-indel, -snp, -substitution) %>% 
    separate(vcf_rowname, into = c("chrom","position","ref","alt")) %>% 
    group_by(chrom, var_type)

set.seed(seed = 531)
random_chroms <- sample(x = unique(denovo_df$chrom),size = 10)
random_df <- data_frame(chrom = random_chroms,
                        var_type = rep(c("indel","snp"), 5))
random_df2 <- data_frame(chrom = random_chroms,
                        var_type = rep(c("snp","indel"), 5))

set_1 <- denovo_df %>% 
    right_join(random_df) %>% 
    group_by(chrom) %>% 
    sample_n(1) %>% 
    dplyr::select(chrom, position, var_type, ref, alt)

set_2 <- denovo_df %>% 
    right_join(random_df2) %>% 
    group_by(chrom) %>% 
    sample_n(1) %>% 
    dplyr::select(chrom, position, var_type, ref, alt)

bind_rows(set_1, set_2) %>% DT::datatable()
```

2. JZ, JW, and NDO manual curation
    - capture image with informative alignment information
    - note reason for classification
3. Summarize curation results with image, classification, and reasoning

# Benchmarking Results
precisionFDA Analysis Note
https://precision.fda.gov/notes/362-gatk4-giab-aj-and-asian-trio-benchmarking 

__Load and tidy data__
```{r}
loadd(benchmark_summary_tbl)

supplemental_table <- benchmark_summary_tbl %>% 
    arrange(Type, Subset) %>% 
    mutate(Subset = if_else(Subset == "*", "all", Subset)) %>% 
    mutate(one_minus_recall = (1 - METRIC.Recall) * 100, 
           one_minus_precision = (1 - METRIC.Precision) * 100) %>% 
    dplyr::select(hgref, Type, Subset, one_minus_recall, one_minus_precision,  
           METRIC.Recall,  METRIC.Precision, METRIC.Frac_NA)

colnames(supplemental_table) <- c("Genome", "Type", "Subset", 
                                 "100% - recall", "100% - precision", 
                                 "Recall", "Precision", 
                                 "Fraction of calls outside high-conf bed")

## Saving to googlesheet for integration paper
int_gs <- gs_title("integrationV3.3.2_supplemental_table2")
gs_edit_cells(ss = int_gs, ws = 1, input = supplemental_table,col_names = TRUE)
```


## Benchmarking Results Table (Supplemental Table 2)
```{r}
supplemental_table %>% DT::datatable()
```



# System Information

```{r}
s_info <- devtools::session_info()
print(s_info$platform)
s_info$packages %>% filter(`*` == "*") %>% dplyr::select(-`*`) %>% 
      knitr::kable()
```

## Software Versions
__BEDtools__
```{r}
system("bedtools --version",intern = TRUE)
```

__bgzip__
```{r}
system("bgzip --version",intern = TRUE)
```


__tabix__
```{r}
system("tabix --version",intern = TRUE)
```

__Real-Time Genomics__
```{r}
system("~/sw/rtg-core-non-commercial-3.9.1/rtg --version",intern = TRUE)[1:2]
```

